spec/edpm-openstack-service-architecture.md

```diff
@@ -0,0 +1,323 @@
+# OpenStack service deployment architecture on EDPM Node
+
+## Problem description
+
+Various OpenStack componentes have agents that need to run on the External
+DataPlane Node. For example OpenStack Nova's nova-api, nova-conductor
+and nova-scheduler services can run in kubernetes pods independently from the
+hypervisors they manage. In the other hand the nova-compute service is designed
+to be collocated with the hyperviors it manages. In the current design the EDPM
+Node serves as a hypervisor for Nova, therefore the nova-compute service needs
+to be deployed on EDPM nodes.
+
+Similar examples from other OpenStack compontents are ovn-metadata-agent, neutron-nicswitch agent, and in the HCI case the cinder-volume service.
+
+### The deployment dependency
+
+The deployment of an EDPM node needs to consider multiple phases
+
+* **phase -1: Bare metal provisioning**. This step transforms an empty bare
+  metal machine to a runnign server with a host OS.
+* **phase 0: Infrastructure deployment and configuration** of storage,
+  networking, and computing resources. This step transforms a generic running
+  server to a more specific storage node, networking node, computing
+  hypervisor, or a mix of these.
+* **phase 1: OpenStack service deployment**. This transforms the server to act
+  as an OpenStack storage node, networking node, or compute node (or a mix of
+  them) by deploying and configuring the specific OpenStack services that
+  connects the server to the OpenStack control plane.
+
+Both phase -1 and 0 can be performed independently from the rest of the
+OpenStack control plane deployment. However phase 1 needs to be coordinated
+with the podified infrastructure, and OpenStack service deployment. For example
+to generate the configuration of the nova-compute service on an EDPM node the
+deployment engine needs to know how to connect the nova-compute service to the
+podified infrasture, e.g. to which nova cell and therefore via which message
+bus. Handling this dependency is one of the scope of this specification.
+
+### The OpenStack service configuration interface
+
+Independently from the fact that an OpenStack service is running in a
+kubernetes pod or in a container on the EDPM Node the service has a well
+defined and widely known configuration interface implemented by oslo.config
+configuration options in one or more *.ini like* files.
+
+The configuration options can be divided to two major parts:
+* configuration that is generated by the deployment engine. E.g. database
+  hostnames and message bus transport URLs
+* configuration that is provided by the human deployer. E.g. CPU allocation
+  ratio, PCI device spec and alias
+
+The existing podified control plane API, the OpenStackControlPlane CRD and the
+service sub CRDs, provides a simple way for the human deployer to pass
+configuration parameters to the OpenStack services. Each CRD that represents
```

Comment:
How I see it. We created k8s CRDs to model OpenStack services. Some of those services running as k8s Pods in the background some of those services running as podman containers on baremetal. The common denominator for me is that both representing OpenStack services not the technology implementing the execution of those services.

For some OpenStack services we will support NodeSelector (e.g. NovaAPI) and therefore those CRDs will have a corresponding field.  For others we wont (e.g. NovaExternalCompute) and therefore there won't be NodeSelector on those CRDs. But still both set of OpenStack services has the same oslo.config configuration interface so both services could be configured via the same oslo.config snippet via the CustomServiceConfig CRD parameter. For me this is a high level and consistent user interface as the use don't have to know which service is executed with witch technology when he or she wants to pass some configuration to it.

> We have plans to expose pass-through k8s options like podTemplate spec

Could you point me to the description of these plans?
The NodeSelector from the podTemplate does not make sense OpenStack services running on the EDPM node, so we don't expose (cannot expose) such configuration on those CRDs. Similarly if another podTemplate parameter does not make sense for OpenStack service on EDPM node then we won't expose them. But for others podTemplate parameters  (e.g. like restartPolicy) that make sense on EDPM node too, we can expose them and implement them.

Reply:
i also dont thnkthat there is agreement aht we should expose

podTemplate, or other k8s constructs in the the CRDs.

we are developing an installer that uses Ansible and k8s internally to deploy OpenStack.
That does not mean we should expose all the options of either technology in our public API.

the pod template in partalare i feel would be incorrect to expose as that somewhat defeats the reason why we are developing operators in the first place.
we are trying to provide a simplified abstraction of the detail of the deployed service
exposing a podTemplate while it would simplify some aspect complicates others especially our support and mantaince of the deploy OpenStack.

Reply:
We have talked about that topic in the past in slack, and something we planned to look at in the next Q. 

I think we need to consider:
- In case of environment specifics or a bug, how can a user customize a deployment without the need to provide a new operator?
- how can support change settings to debug issues?
- have a generic way to customize e.g. timeouts [1]. for routes you use annotations [2] for this.

A generic way e.g. the rabbit operator, we consume, provides is to be able to override annotation and parts of the deployment spec of the service and StatefulSet. That's why we can consume it, can override the service spec to use a metallb loadbalancer, override the deployment spec with our needs. (**Note** it must not be the  full spec as it is a copy where everything is marked as optional). Using this pattern has the advantage to not have to add new parameters for every subset we realize at some point a user need to be able to customize. And we can still limit which parts. Like start with minimum required, like annotations. even our top level replicas could be then an override from the default.

[1] https://issues.redhat.com/browse/OSP-21715
[2] https://docs.openshift.com/container-platform/4.13/networking/routes/route-configuration.html#nw-configuring-route-timeouts_route-configuration
[3] https://github.com/rabbitmq/cluster-operator/blob/main/api/v1beta1/rabbitmqcluster_types.go#L164-L191

Reply:
Can we create a separate discussion for "How to expose podTemplate specific configuration for OpenStack service CRDs running in k8s pods?" I think my proposal quoted below shows that in the discussion of this PR we only need to care about that we *can* expose podTemplate specific parameters on those CRDs if we want. While not exposing them on service CRDs that doesn't or cannot support such parameters.:

> The NodeSelector from the podTemplate does not make sense OpenStack services running on the EDPM node, so we don't expose (cannot expose) such configuration on those CRDs. Similarly if another podTemplate parameter does not make sense for OpenStack service on EDPM node then we won't expose them. But for others podTemplate parameters (e.g. like restartPolicy) that make sense on EDPM node too, we can expose them and implement them.

I do have opinion about exposing the podTemplate directly on OpenStack service CRDs (so please ping me with that discussion if you start a separate one) but I think one way to help the EDPM architectural decision to move forward is to try to split out smaller independent decisions where possible.

Reply:
I've added a paragraph to emphasize that this spec does not try to prescribe that OpenStack service CRDs needs to be aligned regarding every field, e.g. NodeSelector (or podTemplate) might not be exposed for each service CRD.

Reply:
I think it is more than just documenting it in the spec, it is much larger than that. My main concern would be that this is an attempt to represent "external" resources on an entirely separate platform within a CRD that was designed specifically for k8s. It does not follow. I think it goes so far that I would say if we really intend to do this we should re-invision what the OpenStackControlplane CR represents, potentially removing some services, and these top level parameters which were intended to represent and manage services on k8s.

Reply:
I don't see why you think we cannot represent an externally running service as a k8s CRD. I agree that the externally running service should not be repredented by a k8s Pod. But why cannot be represented by a CRD?

There is a doc about [when to add a CRD](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#should-i-add-a-custom-resource-to-my-kubernetes-cluster). I don't see which point there invalidates our idea to represent an external service as a k8s CRD. Could you explain in more details what do you mean by "It does not follow" ?

In the other hand if we accept you point for a second that we cannot represent an external resource as a CRD then DataPlane, DataPlaneRole, DataPlaneNode and DataPlaneService are also invalid CRDs by your definition. What do you suggest there what should we do instead?

Reply:
If you are OK to represent external resources as CRDs in k8s, and your comment expresses that you are against including those external resource CRDs into the OpenStackControlPlane CRD, then I think there is a misunderstanding. The proposal in this document does not suggest that those external resource CRDs are part of the OpenStackControlPlane CRD. If you can point me which part of this doc causes the misunderstanding then I am happy to furter clarify it in the doc.

Reply:
@gibizer It would help to explicitly call this out as not being part of the plan then. We've already had one service that implemented following the NovaExternalCompute model (Telemetry) show up on OpenStackControlplane. We have had previous conversations with Nova folks also state the intent was to treat it as part of the Controlplane. And I think others are confused as well.

Reply:
@gibizer and to clarify your previous comment it is totally fine to represent external services with CRDs.

Reply:
The NovaExternalCompute is part of the contol plane but will not be part of the Nova cr and tasnsitvily will not be part of the OpenstackControlPlane CR.

The NovaCompute CRD will be part of the NovaCellTemplate which is part of the Nova CR and will be part OpenstackControlPlane CR.

The NovaCompute models a podifed nova-compute agent running on opensthif and will be used for the ironic and fake driver.  The NovaExternalCompute will as it is today be used for modelling nova-compute agents running on OpenstackDataPlaneNodes and be free standing form the OpenstackControlPlane CR. the proposal is to directly correlate it by name to the OpenstackDataPlaneNode i.e. and OpenstackDataPlaneNode called "compute-1" with the approatate annotation will result in the creation of a NovaExternalCompute named "compute-1" i would expect the telemetry operator to do the same for ceilometer-compute

create a ceilometerCompute Cr named compute-1 but that would not be in the OpenstackControlPlane CR